File: 42. Trapping Rain Water-3.py
class Solution:
    def trap(self, height: List[int]) -> int:
        res = 0
        st = []
        for i, h in enumerate(height):
            while st and height[st[-1]] <= h:
                bottom_h = height[st.pop()]
                if not st:
                    break
                left = st[-1]
                dh = min(height[left], h) - bottom_h
                res += dh * (i - left - 1)
            st.append(i)
        return res

File: 42. Trapping Rain Water-3.cpp
class Solution {
public:
  int trap(vector<int> &height) {
    int n = height.size();
    stack<int> st;
    int res = 0;
    for (int i = 0; i < n; ++i) {
      int h = height[i];
      while (!st.empty() && height[st.top()] <= h) {
        if (height[st.top()] < h) {
          int b_h = height[st.top()];
          st.pop();
          if (st.empty())
            break;
          int left = st.top();
          int d_h = min(height[left], h) - b_h;
          res += d_h * (i - left - 1);
        } else
          st.pop();
      }
      st.push(i);
    }
    return res;
  }
};